README.md
# 23Volts modules for VCV Rack

This is a free collection of modules for the [VCVRack](http://vcvrack.com) open source modular synthetizer.  

![23volts](res/img/23v-modules.png)
![23volts](res/img/23v-modules-2.png)

## Cells

Cells is a "cellular automata" sequencer based on the [Conway's game of life](https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life). It generates deterministic sequences using the base algorithm and some variations of it.  

- **TICK** : advances to the next generation. When a cable is patched on the grid and has 1 of its 4 neighbours cells alive, it produces a short pulse of the same length of the input clock. When it has 2 or more cells alive, it produces a gate that stays open until the next tick. 

- **ALGO** : Select the current algorithm with a (0-10v) CV signal, otherwise set through the context menu.

- **SPAWN** : generate a new random shape on the grid.

- **CLEAR** : clear the grid and initial state.

- **RESET** : set the grid to the same state it had when the last spawn was triggered. As the algorithm is deterministic, this will result in a loop.

- **DENSITY** : outputs a 0-10v CV value which represents how many cells are "alive" on the grid. 

- **EOL** : A gate which opens when a *"End Of Life"* situation happens, which can be either a stale state or an infinite loop. 

- **OR** : outputs a gate when any of the patched cable is active.

- **NOT** : outputs a gate when none of the patched cables are active. 

## Morph

Morph combines 8 offset generators with a XY Pad which allows to smoothly interpolate between four different "snapshots" of the offset knob's positions. This is a powerful tool for soundesign and live performances. The offset can output from -10  to +10 volts.  

- **X** : allows CV control of the X axis. 

- **Y** : allows CV Control of the Y axis

Alternatively, clicking the **ASSIGN** allows for direct mapping of any parameters of other VCV rack modules. 

## ClockM8

**ClockM8** is a simple yet versatile polyphonic clock modulator. It takes 1 to 16 clocks in its polyphonic input, and outputs up to 16 divided/multiplicated versions of them. 

The ratio is set by the main knob, which can be (polyphonicaly) modulated with the **mod CV input**. The module is designed to stay in sync even if receiving extreme modulation; furthermore, the allowed ratios can be quantized to the following values for more musical results : 

- **Binary** : all strictly binary values 
- **Triplets** : all multiples of 3 
- **Dotted** : dotted binary values
- **Odd** : all other ratios not present in the other like 5,7,9...

**RESET** : The module uses a downbeat approch, which means that regardless of the division/multiplication ratio, the 1st modulated pulse will always generated on the 1st clock tick received after a reset. Also, when a **TRIG** is received from the **CLOCK IN** when the module was not running, a **RESET** signal is internally generated.   

> **About polyphony**: The number of clocks channels of the polyphonic output is defined by the biggest number of channels from the 3 polyphonic inputs (Clock, Mod, Reset). For example, having 2 different resets with a single clock input it will output 2 clocks with a different reset, the first clock input being copied to the 2nd ClockM8 channel.
   

### Setting up output mode : 

By accessing the context menu with a right click you can choose what kind of pulses **ClockM8**. **Trigger** will produce a really short pulse, and **Gate** will produce a pulse half the lenght of the current ratio. 



## Mem

Mem is a Sample & Hold with an internal memory & noise generator that allows for repetable randomness. The size of the memory can be set from 1 to 16K samples, by accessing the context menu. When the end of the memory is reached, *Mem* will switch to its noise generator.  

## Merge8 + Split8

Merge8 and Split8 are respectively a 8 channels polyphonic merger and a 8 channels splitter, only taking 3HP each.

## Merge4 + Split4

Merge4 and Split4 are respectively a 2x4 channels polyphonic merger and a 2x4 channels splitter, only taking 3HP each.

## SwitchN1

SwitchN1 is a polyphonic sequential switch. It takes up to 16 channels as an input which are routed to a mono output. 

The number of step is defined by the number of channels from the input.

- **RESET** : Switch to channel 0. 

- **UP** : Switch to the next channel.

- **DOWN** : Switch to the previous channel

- **RND** : Switch to a random channel. 

- **CV** : Offset the current step with a -5/+5volts signal

## MIDI-PC

MIDI-PC is a simple utility module aimed at sending and receiving MIDI Program Change message from and to VCV rack.

## Multimap

Multimap is a MIDI mapping tool with feedback which can use the same MIDI-CC or notes and assign them to up to 128 banks of different assignments inside VCV. 

It comes in 2 versions; one version has 16 knobs and one version has 16 switches.

## MonoPoly

Mono poly takes a mono input and copy the signal to N polyphonic channels. The number of channels can be set by CV.

## Polymerge & Polysplit

Polymerge & Polysplit are channel-variable merger & splitter. Right click on the module allows for selecting the number of channels of the inputs / outputs.

## Contact

If you found a bug or would like to suggest a new feature, please open a github issue. In the case you don't have a github account nor want to create one, feel free to send an email at **23volts@gmail.com**

## Donating

If you use these modules and enjoy them, and would like to support their development, you can make a donation through [PayPal](https://www.paypal.me/23volts)  


CHANGELOG.md
### 1.1.5 (2019-02-16)
- PolyMerge : fix ordering of polyphonic inputs

### 1.1.4 (2019-02-05)
- Morph & Multimap : Fix knob glitch when mapped value closes to 0
- Morph : Disable autolearn when X or Y CV input is connected

### 1.1.3 (2019-01-20)
- Morph : Add option to reverse Y AXIS for MIDI control
- Multimap : Fix bank switching's color update when using CV & FIRST inputs
- Multimap : Set ParamHandle transparency on preset load
- SwitchN1 : Using STEP INC input was causing selector to skip 1st step, now fixed.

### 1.1.2 (2019-01-12)
- Morph : fixed crash happening in module's browser

### 1.1.1 (2019-01-11)
- Multimap : Allow chain assignment. 
- Multimap : Mapping indicators are now displayed in transparency if they don't belong to the active bank. 
- Morph : Allow chain assignment.
- Morph & Multimap : Assigning parameters now doesn't "lock" the parameter on the target module.
- Morph : Add MIDI Learn for X & Y axis

### 1.1.0 (2019-01-03)

- Introducing **Multimap**, a versatile MIDI-Map utility with feedback.
- Introducing **MIDI-PC**,  an utility to send/receive MIDI Program Change messages.
- Introducing **MonoPoly**, a mono to poly signal converter.
- Introducing **PolyMerge**, a poly to poly merger.
- Introducing **PolySplit**, a poly to poly splitter.
- Introducing **Split4**, a 2x4 voices signal splitter.
- Introducing **Merge4**, a 2x4 voices signal merger.
- Cells : spawned cells are now more likely to appear in the center of the grid.
- Cells : increase the range of infinite loop detection.
- Switch N1 : Add Reset Input.
- Morph : A snapshot is now automatically written when XY pad is set on it.
- Morph : X + Y CV inputs now don't affect user's selected pad position.
- Morph : Add Assign mode, which allows directly mapping knobs to other VCV modules.
- All modules : skin update + layout & readability enhancements.

### 1.0.1 (2019-11-14)
- Cells : fixes crash that was happening when Algo CV was exactly 0v.
- Cells : fixed Density, now outputs correct value.
- Morph : fixes font path.

### 1.0.0 (2019-11-13)
Introducing **Cells**, **Morph**, **ClockM8**, **Mem**, **Merge8**, **Split8**, **SwitchN1**.

src/Mem.cpp
#include "23volts.hpp"
#include "widgets/ports.hpp"

struct MemoryBank {
	private:
		int m_max_size;
		float* m_data;

		void initData() {
			m_memory = new std::vector<float>;
			m_memory->reserve(m_max_size);
			m_data = m_memory->data();
		}

	public:
		std::vector<float>* m_memory;
		
		MemoryBank(int maxSize) 
		{
			m_max_size = maxSize;
			initData();
		}
		~MemoryBank() 
		{
			delete m_memory;
		}

		float* data() const 
		{
			return m_data;
		}

		int size() const
		{
			return m_memory->size();
		}

		float getValue(int position) const
		{
			return m_data[position];
		}

		void clear() 
		{
			delete m_memory;
			initData();
		}

		/**
		 * Fill the internal data with data from another vector. 
		 * Source if truncated if the size of source vector > m_max_size
		 */
		void fill(std::vector<float>* sourceVector)
		{
			float* data = sourceVector->data();
			for (int x=0; x < (int) sourceVector->size() && x < m_max_size; x++) {
				m_memory->push_back(data[x]);
			}
		}

};

struct PureNoiseSource {
	float getValue() {
		return 2.0 * random::normal();
	}
};

struct StatisticNoiseSource {
	int m_max_size;
	
	MemoryBank *m_memory;
	std::vector<float> noiseTable;

	StatisticNoiseSource(int maxSize) {
		m_memory = new MemoryBank(maxSize);
	}

	~StatisticNoiseSource() {
		delete m_memory;
	}

	float getValue() {
		return 0.f;
	}

	/**
	 * Append memory to stats, after a clear for example, or each x samples
	 */
	void append(MemoryBank &addedMemory) {
		if(addedMemory.size() + m_memory->size() > m_max_size) return;

		// Append memory
	}

	void generateNewNoiseTable(float noiseAmount, int size) {
		// Create a new noise table
	}
};

struct Mem : Module {
	enum ParamIds {
		NUM_PARAMS
	};
	enum InputIds {
		CLEAR_INPUT,
		INPUT_INPUT,
		TRIG_INPUT,
		RESET_INPUT,
		WRITE_INPUT,
		NUM_INPUTS
	};
	enum OutputIds {
		OUTPUT_OUTPUT,
		NUM_OUTPUTS
	};
	enum LightIds {
		WRITE_LIGHT,
		READ_LIGHT,
		NOISE_LIGHT,
		NUM_LIGHTS
	};

	static const int MEMORY_OPTIONS = 23;
	const int MEMORY_SIZES[23] = {1,2,3,4,6,7,8,12,13,16,23,32,42,48,64,96,128,256,1024,2048,4096,8192,16384};
	static const int DEFAULT_MEMORY_SIZE = 16; //_INDEX

	static const int PROCESS_BUFFER_SIZE = 32;
	static const int IDLE_BUFFER_SIZE = 1152000;

	int processBuffer = 0;
	int idleBuffer = 0;

	dsp::SchmittTrigger trigTrigger;
	dsp::SchmittTrigger clearTrigger;
	dsp::SchmittTrigger resetTrigger;

	bool isIdle = true;

	bool isReading = false;
	bool isWriting = false;
	bool isRandomizing = false;

	bool isInputConnected = false;
	bool isWriteConnected = false;

	MemoryBank* memory;
	int memorySize = MEMORY_SIZES[DEFAULT_MEMORY_SIZE];
	float currentOutputValue = 0.f;
	int position = -1;

	PureNoiseSource noiseSource;
	StatisticNoiseSource* statisticSource;

	Mem() {
		config(NUM_PARAMS, NUM_INPUTS, NUM_OUTPUTS, NUM_LIGHTS);

		memory = new MemoryBank(MEMORY_SIZES[DEFAULT_MEMORY_SIZE]);
		statisticSource = new StatisticNoiseSource(MEMORY_SIZES[MEMORY_OPTIONS - 2]);
	}
	
	~Mem() {
		delete memory;
		delete statisticSource;
	}

	void reset() {
		position = -1;
		isReading = false;
		isRandomizing = false;
		isWriting = false;
	}

	void process(const ProcessArgs &args) override {

		if(processBuffer == 0) processTimely(args);
		if(idleBuffer == 0) processIdle(args);

		bool writeGate = isWriteConnected && inputs[WRITE_INPUT].getNormalVoltage(0.f) >= 1.f;

		if (clearTrigger.process(inputs[CLEAR_INPUT].getVoltage())) {
			// TODO Add to stats
			memory->clear();
			reset();
		}

		if (resetTrigger.process(inputs[RESET_INPUT].getVoltage())) {
			reset();
		}

		if (trigTrigger.process(inputs[TRIG_INPUT].getVoltage())) {
			isIdle = false;
			idleBuffer = IDLE_BUFFER_SIZE;
			position++;

			if((memory->size() > position) && ! writeGate) {
				isReading = true;
				isWriting = false;
				isRandomizing = false;
				currentOutputValue = memory->getValue(position);
			}
			else {
				isReading = false;
				if(memorySize > position) {
					isWriting = true;
					// we have still some place left
					if(isInputConnected) {
						currentOutputValue = inputs[INPUT_INPUT].getVoltage();
					}
					else {
						isRandomizing = true;
						currentOutputValue = noiseSource.getValue();
					}
					memory->m_memory->push_back(currentOutputValue);
				}
				else {
					isWriting = false;
					isRandomizing = true;
					currentOutputValue = noiseSource.getValue();
				}
			}
			
		}
		else {
			idleBuffer--;
		}

		lights[WRITE_LIGHT].setBrightness(isWriting ? 1.f : 0.f);
		lights[READ_LIGHT].setBrightness(isReading ? 1.f : 0.f);
		lights[NOISE_LIGHT].setBrightness(isRandomizing && ! isIdle ? 1.f : 0.f);

		outputs[OUTPUT_OUTPUT].setVoltage(currentOutputValue);

		processBuffer--;
	}


	/**
	 * Part of the process only executed every PROCESS_BUFFER samples
	 */
	void processTimely(const ProcessArgs &args) {

		// Check states of connected cables
		isWriteConnected = inputs[WRITE_INPUT].isConnected();
		isInputConnected = inputs[INPUT_INPUT].isConnected();

		// swap noise statistics taken 

		processBuffer = PROCESS_BUFFER_SIZE;
	}

	/**
	 * Process some time consuming stuff, like writing to disk, when no clock has
	 * been received for a long time (>10-20s)
	 */
	void processIdle(const ProcessArgs &args) {

		idleBuffer = IDLE_BUFFER_SIZE;
		isIdle = true;
	}


	json_t *dataToJson() override {
		json_t *rootJ = json_object();

		json_object_set_new(rootJ, "outputValue", json_real(currentOutputValue));
		json_object_set_new(rootJ, "position", json_integer(position));
		
		uint8_t *memoryData;
		memoryData = reinterpret_cast<uint8_t*>(memory->data());
		int memoryByteSize = memory->size() * 8;
		std::string memoryStr = string::toBase64(memoryData, memoryByteSize);
		json_object_set_new(rootJ, "memory", json_string(memoryStr.c_str()));
		json_object_set_new(rootJ, "memorySize", json_integer(memorySize));
		json_object_set_new(rootJ, "memoryByteSize", json_integer(memoryByteSize));

		json_object_set_new(rootJ, "isWriting", json_boolean(isWriting));
		json_object_set_new(rootJ, "isReading", json_boolean(isReading));
		json_object_set_new(rootJ, "isRandomizing", json_boolean(isReading));

		return rootJ;
	}

	void dataFromJson(json_t *rootJ) override {

		currentOutputValue = json_real_value(json_object_get(rootJ, "outputValue"));
		position = json_integer_value(json_object_get(rootJ, "position"));
		memorySize = json_integer_value(json_object_get(rootJ, "memorySize"));

		// memory
		unsigned long memoryByteSize = json_integer_value(json_object_get(rootJ, "memoryByteSize"));
		std::string memoryStr = json_string_value(json_object_get(rootJ, "memory"));
		std::vector<uint8_t> memoryVector = string::fromBase64(memoryStr);
		float* actualData = reinterpret_cast<float*>(memoryVector.data());

		memory = new MemoryBank(memorySize);
		for(unsigned long x = 0; x < memoryByteSize / 8; x++) {
			memory->m_memory->push_back(actualData[x]);
		}

		isWriting = json_is_true(json_object_get(rootJ, "isWriting"));
		isReading = json_is_true(json_object_get(rootJ, "isReading"));
		isRandomizing = json_is_true(json_object_get(rootJ, "isRandomizing"));
	}

	void setMemorySize(int size) {
		MemoryBank* newMemory = new MemoryBank(size);
		newMemory->fill(memory->m_memory);
		delete memory;
		memory = newMemory;
		memorySize = size;
	}
};

struct MemorySizeValueItem : MenuItem {
	int m_size;
	Mem* module;

	MemorySizeValueItem(int size) {
		m_size = size;
	}

	void onAction(const event::Action& e) override {
		module->setMemorySize(m_size);
	}
};

struct MemWidget : ModuleWidget {

	MemWidget(Mem *module) {
		setModule(module);
		setPanel(APP->window->loadSvg(asset::plugin(pluginInstance, "res/Mem.svg")));

		addChild(createWidget<ScrewSilver>(Vec(RACK_GRID_WIDTH, 0)));
		addChild(createWidget<ScrewSilver>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, 0)));
		addChild(createWidget<ScrewSilver>(Vec(RACK_GRID_WIDTH, RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));
		addChild(createWidget<ScrewSilver>(Vec(box.size.x - 2 * RACK_GRID_WIDTH, RACK_GRID_HEIGHT - RACK_GRID_WIDTH)));

		addInput(createInputCentered<PJ301MPort>(mm2px(Vec(7.7, 16.72)), module, Mem::TRIG_INPUT));
		addInput(createInputCentered<SmallPort>(mm2px(Vec(7.7, 62.758)), module, Mem::INPUT_INPUT));
		addInput(createInputCentered<SmallPort>(mm2px(Vec(7.7, 77.234)), module, Mem::CLEAR_INPUT));
		addInput(createInputCentered<SmallPort>(mm2px(Vec(7.7, 91.862)), module, Mem::RESET_INPUT));
		addInput(createInputCentered<InvisiblePort>(mm2px(Vec(10.69, 51.23)), module, Mem::WRITE_INPUT));

		addOutput(createOutputCentered<PJ301MPort>(mm2px(Vec(7.7, 107.503)), module, Mem::OUTPUT_OUTPUT));

		addChild(createLight<MediumLight<GreenLight>>(mm2px(Vec(10, 36.6)), module, Mem::READ_LIGHT));
		addChild(createLight<MediumLight<RedLight>>(mm2px(Vec(10, 40.4)), module, Mem::WRITE_LIGHT));
		addChild(createLight<MediumLight<BlueLight>>(mm2px(Vec(10, 44.2)), module, Mem::NOISE_LIGHT));
	}

	void appendContextMenu(Menu* menu) override {
		Mem* module = dynamic_cast<Mem*>(this->module);

		menu->addChild(new MenuSeparator);

		// TODO Context menu -> add "Enable noise statistics"
		
		MenuLabel* item = new MenuLabel;
	 	item->text = "Memory Size";
	 	menu->addChild(item);

		// Context menu -> add memory size select list
		for(int x = 0; x < module->MEMORY_OPTIONS; x++) {
			MemorySizeValueItem* menuItem = new MemorySizeValueItem(module->MEMORY_SIZES[x]);
			std::string label = std::to_string(module->MEMORY_SIZES[x]) + " samples" ;
			menuItem->text = label;
			menuItem->module = module;
			menuItem->rightText = CHECKMARK(module->MEMORY_SIZES[x] == module->memorySize);
			menu->addChild(menuItem);
		}

	}
};

Model *modelMem = createModel<Mem, MemWidget>("Mem");
